design/baremetal-operator/vendor-firmware-support.md

```diff
@@ -0,0 +1,151 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Add API to support vendor firmware burn and config
+
+## Status
+
+provisional
+
+## Summary
+
+Add a new interface to allow burn and config vendor firmware.
+
+## Motivation
+
+It desired to burn and config vendor firmware before deploying the OS image.
+With hardware manager framework vendor can extend IPA to burn and configure its firmware.
+Burning and configuring firmware is done in the manual cleaning step. It desired to expose this
+functionality to BMO API.
+
+### Goals
+
+1. Agree on the format of the API and attributes
+2. Agree on Nvidia vendor format
+
+### Non-Goals
+
+1. Agree on others vendor’s format.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+Cloud admin wants to burn and configure vendor firmware before OS deployed on bare-metal host.
+
+## Design Details
+
+### Implementation Details/Notes/Constraints
+
+
+IPA HardwareManagers allows vendors to add firmware burn and config
+in the manual cleaning stage.
+
+The ironic command to trigger the manual cleaning would be in
+the following syntax:
+
+```shell
+baremetal node clean node --clean-steps \
+[{"interface": "deploy", "step": <vendor function>, \
+   "args": {<vendor_arg1>: <vendor_value1>, \
+   <vendor_arg2>: <vendor_value2>, ...}}]
+```
+
+IPA HardwareManagers framework defines the following attributes:
+1. `step` - A function on the vendor hardware manager
+2. `argsinfo` - Arguments that can be required or optional to the step function
+3. `interface` - Should always be the deploy interface
+
+The proposal is to extend the BareMetalHost with VendorFirmware field.
+The VendorFirmware field includes the `step` name and `args` that can be required or optional.
+
+Nvidia hardware manager support `step` - update_nvidia_firmware with `args` - firmware_config and firmware_url
+
+The VendorFirmwareConfig for Nvidia should be the following:
+
+```yaml
+vendor_firmware:
```

Comment:
an alternative is to model clean steps into BMH.  just a thought here...

that is something of this sort:

```
cleanSteps:
  - interface : "deploy"
    step: "update_nvidia_firmware"
    args:
      firmware_config: "http://10.7.12.161/07_2022/config.yaml"
      firmware_url: "http://10.7.12.161/fw/07_2022"
  - interface: "deploy"
    step "another_step"
    args:
        other: "val1"
        yetAnother: "val2"
```


args in this case would be a `map[string] string`

Reply:
I think I cover this on the Alternatives section [1]

[1] -  https://github.com/metal3-io/metal3-docs/pull/290/files/fc5600f7ed3075a649f331ab06887dec07e4c755#diff-9e85ae0576b8227ae061d315a315390d1c6fbfd9bd3cdd7ca27ace7d94f5f69cR128

Reply:
yes, its similar, but what i propose above is a bit more structured imo.

Reply:
Ironic is very explicitly a low-level system, while we have been so far trying to keep BMO API high-level. Exposes clean steps as they are raises a lot of questions. For example: which steps have to run every time and which - only once.